### PR TITLE
Download, expose and use Chart Readme files

### PR DESCRIPTION
**Pull request by migmartri**
## TLDR:

* Backend: Fetches and exposes specific chartVersion Readme.md files via the API https://github.com/helm/monocular/issues/31
* Frontend: Uses this new readme file https://github.com/helm/monocular/issues/32

## Importation
* During repository fetching, we download and store in the filesystem ($HOME/repo-data) the Readme files for those versions.  We must persist that directory as a volume, otherwise subsequent `refresh` calls will be expensive.

* This mechanism can be useful if we decide to store icons on importation time so we can process them and avoid possible issues related to hotlinking https://github.com/helm/monocular/issues/43 
selection_092

## Exposure
* Readme files are exposed via `.../version/{version}/readme` API endpoint following open-api.org convention
selection_096
selection_097

Closes https://github.com/helm/monocular/issues/31

## UI

selection_098

Closes  https://github.com/helm/monocular/issues/32

## Caveats

* I know that testing and coverage could be improved but I'd prefer to land this asap so I do not block the work happening in the UI.
* I'm not familiar with Go so I tried to follow existing code conventions, in any case it is likely that I am making basic mistakes so harsh feedback is welcome :)

### CODE REVIEW

File: src/api/data/cache/cache.go [hunk 1]
```diff
@@ -143,6 +144,13 @@ func (c *cachedCharts) Refresh() error {
 			c.allCharts[repoName] = []*models.ChartPackage{}
 			for _, chart := range charts {
 				chart.Repo = repoName
+				// Extra files. Skipped if the directory exists
+				if _, err := os.Stat(chartDataDir(chart)); os.IsNotExist(err) {
```
Comment (migmartri) [inline src/api/data/cache/cache.go]: We download the files if cache does not exist for that specific version.

File: src/api/handlers/charts.go [hunk 2]
```diff
@@ -41,6 +42,28 @@ func GetChartVersion(params operations.GetChartVersionParams, c data.Charts) mid
...
+func GetChartVersionReadme(params operations.GetChartVersionReadmeParams, c data.Charts) middleware.Responder {
+	// We check that the chart exists before load the readme
+	chartPackage, err := c.ChartVersionFromRepo(params.Repo, params.ChartName, params.Version)
+	if err != nil {
+		log.Printf("data.Charts.ChartVersionFromRepo(%s, %s, %s) error (%s)", params.Repo, params.ChartName, params.Version, err)
+		return notFound(ChartVersionResourceName)
+	}
+
+	filename := "README.md"
+	readmeContent, err := cache.ReadFromCache(chartPackage, filename)
```
Comment (migmartri) [inline src/api/handlers/charts.go]: Using the convenience method created in `cache package`

File: src/api/swagger.yml [hunk 3]
```diff
@@ -88,6 +88,32 @@ paths:
...
+          required: true
+        - name: version
+          in: path
+          type: string
+          required: true
+      responses:
+        200:
+          description: Readme.md file content
+          schema:
+            $ref: "#/definitions/resourceDataString"
```
Comment (migmartri) [inline src/api/swagger.yml]: New resource to simplify response. It includes `data` root. I need to add `type` and `id` to the payload to make it fully complaint with the jsonapi spec

**Comment by Angelmmiguel** [comment]
Are we going to parse and render the README in Frontend @migmartri??

**Comment by migmartri** [comment]
> Are we going to parse and render the README in Frontend @migmartri??
@Angelmmiguel, retrieval, parsing and rendering will be UI concern. 

We will need to remove the mocked Readme file we have in the UI and make a call to the provided version specific API endpoint instead.

File: src/api/data/cache/cache-helper.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
+package cache
```
Comment (prydonius) [inline src/api/data/cache/cache-helper.go]: This filename isn't really idiomatic, I suggest dropping the cache prefix and going with `helpers.go` or `tar_helpers.go` or `cache_helper.go`.

File: src/api/handlers/charts.go (hunk 2 above)
Comment (technosophos) [inline src/api/handlers/charts.go]: That looks good.

File: src/api/data/cache/cache-helper.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+		return err
+	}
+
+	if !tarballExists(chart) {
+		if err := downloadTarball(chart); err != nil {
+			return err
+		}
+	}
+
+	if err := extractFilesFromTarball(chart); err != nil {
```
Comment (technosophos) [inline src/api/data/cache/cache-helper.go]: tiny nit: you could just simplify this to `return extractFilesFromTarball(chart)`

File: src/api/data/cache/cache-helper.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+	fmt.Printf("Downloading metadata from %s\n", source)
+	// Download tarball
+	resp, err := http.Get(source)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
```
Comment (technosophos) [inline src/api/data/cache/cache-helper.go]: similar, here you could just `return err`.
